internal/app/appauth/storage.go

```diff
@@ -0,0 +1,347 @@
+package appauth
+
+import (
+	"context"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"errors"
+	"math/big"
+	mrand "math/rand"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/caos/oidc/pkg/oidc"
+	"github.com/caos/oidc/pkg/op"
+	"github.com/oklog/ulid"
+	"gopkg.in/square/go-jose.v2"
+)
+
+type Storage struct {
+	lock      sync.Mutex
+	appConfig *StorageConfig
+	clients   map[string]op.Client
+	requests  map[string]AuthRequest
+	keySet    jose.JSONWebKeySet
+	key       *rsa.PrivateKey
+}
+
+type StorageConfig struct {
+	Domain string `default:"http://localhost:8080"`
+	Debug  bool
+}
+
+func NewAuthStorage(cfg *StorageConfig) op.Storage {
+
+	s := &Storage{
+		appConfig: cfg,
+	}
+
+	initData(s)
+	initKeys(s)
+
+	return s
+}
+
+func initData(s *Storage) {
+
+	client := initLocalClient(s.appConfig.Debug)
+
+	s.clients = map[string]op.Client{
+		client.GetID(): client,
+	}
+
+	s.requests = make(map[string]AuthRequest)
+}
+
+func initKeys(s *Storage) {
```

Comment:
```suggestion
func initKeys(name pkix.Name) (*rsa.PrivateKey, jose.JSONWebKeySet) {
```

The code that causes the side effect may reduce the versatility, maintainability, and testability of the function more than necessary.

In this case, it should be the responsibility of the caller to set the key to the storage. By lines of `s.xxx = xxx`, the versatility, maintainability, and testability of initKeys function is lost.